### PR TITLE
MDEV-15543: tmpfile.d not for datadir

### PR DESCRIPTION
**Pull request by grooverdan**
- [x] *The Jira issue number for this PR is: MDEV-15543*

## Description

With the default datadir path of /var/lib/mysql the same as the path of the unix socket, this tmpfiles.d file was generated to "create" a datadir.

Don't do this if these configuration items are the same.

## How can this PR be tested?

Check if file is created and included in RPMs.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [X] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [X] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/11.0/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [X] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

**Review by cvicentiu — COMMENTED** [review]
Hi @grooverdan!

In general I do not agree with this patch because it feels like it's adding a band-aid over a problem that is created somewhere else (our `install_layout.cmake` RPM config), but I am not 100% sure I understand the problem we are fixing. Let me explain my reasoning, then you can comment if I got something wrong:

Our default RPM packaging uses:
```cmake
SET(INSTALL_MYSQLDATADIR_RPM            "/var/lib/mysql")
SET(INSTALL_UNIX_ADDRDIR_RPM            "${INSTALL_MYSQLDATADIR_RPM}/mysql.sock")
```
When building RPM packages,
` MYSQL_UNIX_ADDR` a bit later down this CMakeLists.txt file takes the same value as `INSTALL_UNIX_ADDRDIR_RPM`

So in the end we're left with:

```c
MYSQL_DATADIR="/var/lib/mysql"
MYSQL_UNIX_ADDR="/var/lib/mysql/mysql.sock"
 ```
 These are our default settings, which can be configured differently at cmake time, but I don't want to cover that aspect because that's for "advanced users" that know what they're doing when compiling the server.
 
 Note that I am not questioning whether the MYSQL_DATADIR and MYSQL_UNIX_ADDR values are actually correct. I personally wouldn't have expected the socket to be in the same place as the data directory, (rather something like `/run/mariadb/` or something such) but it's been like this since 2012. If we are to change it, it should be in a newer version of MariaDB.
 
 Now, what the user in MDEV-15543 is complaining about is:
 
 1. The user installed our RPMs with the default settings we use to compile them.
 2. The user complains that we shipped /usr/lib/tmpfiles.d/tmpfiles.conf.
 
I think the user is wrong here, we should have shipped /usr/lib/tmpfiles.d/mariadb.conf, because of the "RENAME mariadb.conf" directive in CMake. So probably the MDEV text has a typo there. Nonetheless, I'll consider that the user is refering to the mariadb.conf tmpfiles specification file. The contents of that file, as is generated by our `CONFIGURE_FILE` step will be:

```
d @MYSQL_UNIX_DIR@ 0755 @MYSQLD_USER@ @MYSQLD_USER@ -
```
Here, `MYSQL_UNIX_DIR` is `/var/lib/mysql/` and `MYSQLD_USER` is mysql.

So we get to the contents:

```
d /var/lib/mysql 0755 mysql mysql -
```

Note that all this is done at compile time, before the package is installed by the user.

What the user now does is:

3. Changes the default data directory of the server to some other path (say `/mnt/database/`, completely unrelated to `/var/lib/mysql`).
3a. Changes the default unix socket directory from `/var/lib/mysql/mysql.socket` to something else, such that `/var/lib/mysql/` is now just an empty dir.
4. Presumably runs mariadb-install-db script on the new datadir and starts the server.
5. The user then sees that systemd respects tmpfiles.conf for mariadb and creates what would be the default value for the unix socket directory: `/var/lib/mysql`.

My opinion is that this is not something that would get fixed by this patch because this is an issue created by not doing a complete server configuration change on the user's server and the actual tmpfiles.conf files have already been set in the rpm package.

What the user should have done, should he want to change his default datadir and also not have the issue with systemd-tmpfiles:

a. change the contents of /usr/lib/tmpfiles.d/mariadb.conf to point to the right UNIX socket directory.
or
b. delete the /usr/lib/tmpfiles.d/mariadb.conf file entirely.

You patch has the same effect as b but do we really actually need this? Any deviation from default settings (say a user changes the socket dir) makes this systemd-tmpfiles entry useless, be it for debian, redhat or our standalone install.

So, my suggestion would be to either:
1. Not do anything and close this as not a bug. (which is where I lean towards the most)

or

2. Remove the `INSTALL_SYSTEMD_TMPFILESDIR_RPM` variable entirely.
This will have the same effect as your patch, for our default CPack setup, but it will impact distros that do change those values during their compilation step. Then again, I suspect such distros actually will run cmake with:
`-DINSTALL_SYSTEMD_TMPFILESDIR=/usr/lib/tmpfiles.d` if they do want the tmpfiles anyway.

or

3. Keep your change, but add a comment explaining why it's necessary.
I'm not in favor of this because it doesn't really improve or simplify the situation. It just hacks a fix for an otherwise rather generic and clean tmpfiles.conf install rule. I'm also unsure if suddenly not shipping a file in a stable MariaDB is appropriate. Perhaps users actually relied on `/var/lib/mysql/` being created specifically for the socket file.
We'd be breaking their installs when they upgrade to the latest 10.4.

**Comment by grooverdan** [comment]
@cvicentiu thanks for the review.

The user complain wasn't about the name `/usr/lib/tmpfiles.d/tmpfiles.conf` (MDEV-17028 was, and was written and fixed after the bug).

The essence of the bug was that the datadir was in the tmpfile.d file, however named. And tmpfiles shouldn't contain at least the default datadir.

User changes can happen. I'm not sure what you mean by "5. The user then sees that systemd...". This is either:
a) user sees this and changes this to their socket dir, probably ok
b) user sees this and changes to their datadir, probably not ok for the purposes of the bug.
c) user sees this and wonders why `/var/lib/mysql` keeps reappearing.

The most negative aspect of this change is with a), that they may not know tmpfiles.d can create temporary directories like `/run/mariadb`.

So this patch preserves the creation of a `tmpfile.d/mariadb.conf` for when INSTALL_UNIX_ADDRDIR set by a distro isn't the datadir, mainly aimed at helping out the default configuration of those distros.

I'd rather protect users with a default datadir from an parser of the file that considers /var/lib/mysql temporary than users actually relying on /var/lib/mysql/ being created specifically for the socket file. But maybe if we haven't had complaints we could do this for 11.3/11.4?

**Review by cvicentiu — COMMENTED** [review]
@grooverdan I just had an idea:

Why not keep shipping the file, but have the path inside it  commented out in case the datadir matches the tmpfiles dir?

We could also add an extra comment on top to explain what the file is for? Just like we offer user suggestions in my.cnf.

**Comment by grooverdan** [comment]
> @grooverdan I just had an idea:
> 
> Why not keep shipping the file, but have the path inside it commented out in case the datadir matches the tmpfiles dir?
> 
> We could also add an extra comment on top to explain what the file is for? Just like we offer user suggestions in my.cnf.

I've usually found my.cnf suggestions bloated, but it could work here as suggested. Sure.

File: support-files/tmpfiles.conf.in [hunk 1]
```diff
@@ -1 +1,11 @@
-d @MYSQL_UNIX_DIR@ 0755 @MYSQLD_USER@ @MYSQLD_USER@ -
+# This is the directory where the unix socket
+# of MariaDB is created.
+# Other temporary directies can be created here like:
```
Comment (cvicentiu) [inline support-files/tmpfiles.conf.in:3]: typo: directories

**Review by cvicentiu — APPROVED** [review]
OK to push with typo fix.

**Pull request merged**